yxi/xml/xml_generator.py

```diff
@@ -0,0 +1,113 @@
+import sys
+import json
+from xml.etree.ElementTree import Element, SubElement, tostring
+from xml.dom import minidom
+from math import log2
+
+"""
+This file takes in a `.yxi` description and outputs a xml suitable for a `kernel.xml` file
+can be used to package an xclbin for the Xilinx XRT runtime.
+See https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File
+for the spec this is based on.
+"""
+size_key = "total_size"
+width_key = "data_width"
+
+def gen_xml(yxi):
+  mems = yxi["memories"]
+  check_mems_wellformed(mems)
+
+  root = Element("root", {"versionMajor": "1", "versionMinor": "6"})
+  kernel = SubElement(root, "kernel", {
+    "name" : yxi["toplevel"],
+    "language": "ip_c",
+    # TODO: Make sure this matches component.xml, Namely the `Toplevel` part.
+    #  See https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File
```

Comment:
In particular, I think `Toplevel` here is the name of the Verilog module in question?

(I'm not 100% sure what you mean by matching `component.xml`, since we are not providing that file.)

Reply:
Right, the old system created a toplevel `Toplevel` module, while the new system creates a toplevel `wrapper` module. This could easily be changed, although perhaps "[AXI-]wrapper"  is a bit more descriptive as to what the toplevel module is doing.

Regarding matching the `component.xml` that is a requirement from here:

```
Must match the vendor, library, name, and version attributes in the `component.xml` of an IP. For example, if component.xml has the following tags:
<spirit:vendor>xilinx.com</spirit:vendor>

<spirit:library>hls</spirit:library>

<spirit:name>test_sincos</spirit:name>

<spirit:version>1.0</spirit:version>

The vlnv attribute in kernel XML must be set to:xilinx.com:hls:test_sincos:1.0
```

I'm actually a bit confused as to where the `component.xml` is generated/specified. It seems like it is [this line](https://github.com/calyxir/calyx/blob/efffa92ec94c0e577ea7c88db5cf7b874893af7b/fud2/rsrc/gen_xo.tcl#L27)
has information regarding the above tags. However the [docs](https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File) more often refer to Vitis being able to generate a `kernel.xml` from a `component.xml`, and doesn't seem to reference going in the opposite direction. But I imagine that we can if that's what we've been doing until now?

Reply:
Yeah, I could be wrong, but I _believe_ that `component.xml` is created for us in the `package_xo` step. And it does that using the `kernel.xml` that we provide. So presumably all that matters here is that we provide the name of our top-level Verilog module.